File: products/LoanHub/apps/backend/database/schemas/company_operating_system.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, HttpUrl


CompanyModule = Literal[
    "crm", "credit_committee", "collateral", "legal_recovery", "complaints",
    "communications", "marketing", "agents", "employer_partnerships",
    "procurement", "assets", "internal_audit", "budgeting", "targets",
    "business_continuity", "integrations", "document_automation", "board_packs",
]


class OperatingRecordCreate(BaseModel):
    module: CompanyModule
    record_type: str = Field(min_length=2, max_length=80)
    title: str = Field(min_length=2, max_length=240)
    description: str | None = Field(default=None, max_length=5000)
    status: str = Field(default="open", min_length=2, max_length=40)
    priority: str = Field(default="normal", min_length=2, max_length=30)
    branch_id: UUID | None = None
    borrower_id: UUID | None = None
    loan_id: UUID | None = None
    assigned_user_id: UUID | None = None
    counterparty_name: str | None = Field(default=None, max_length=240)
    amount: Decimal | None = None
    currency: str = Field(default="LSL", min_length=3, max_length=3)
    due_at: datetime | None = None
    data: dict[str, Any] = Field(default_factory=dict)
    tags: list[str] = Field(default_factory=list, max_length=30)


class OperatingRecordUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=2, max_length=240)
    description: str | None = Field(default=None, max_length=5000)
    status: str | None = Field(default=None, min_length=2, max_length=40)
    priority: str | None = Field(default=None, min_length=2, max_length=30)
    assigned_user_id: UUID | None = None
    counterparty_name: str | None = Field(default=None, max_length=240)
    amount: Decimal | None = None
    due_at: datetime | None = None
    data: dict[str, Any] | None = None
    tags: list[str] | None = Field(default=None, max_length=30)
    is_archived: bool | None = None


class OperatingRecordRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    company_id: UUID
    branch_id: UUID | None
    module: str
    record_type: str
    reference: str
    title: str
    description: str | None
    status: str
    priority: str
    borrower_id: UUID | None
    loan_id: UUID | None
    assigned_user_id: UUID | None
    counterparty_name: str | None
    amount: Decimal | None
    currency: str
    due_at: datetime | None
    data: dict[str, Any]
    tags: list[str]
    is_archived: bool
    created_at: datetime
    updated_at: datetime


class APIKeyCreate(BaseModel):
    name: str = Field(min_length=2, max_length=160)
    scopes: list[str] = Field(default_factory=lambda: ["read:portfolio"], max_length=30)
    allowed_ips: list[str] = Field(default_factory=list, max_length=30)
    expires_at: datetime | None = None


class APIKeyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    name: str
    key_prefix: str
    scopes: list[str]
    allowed_ips: list[str]
    expires_at: datetime | None
    last_used_at: datetime | None
    revoked_at: datetime | None
    created_at: datetime


class APIKeyIssued(APIKeyRead):
    api_key: str


class WebhookCreate(BaseModel):
    name: str = Field(min_length=2, max_length=160)
    endpoint_url: HttpUrl
    event_types: list[str] = Field(default_factory=lambda: ["loan.updated", "payment.succeeded"], max_length=30)


class WebhookRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    name: str
    endpoint_url: str
    secret_prefix: str
    event_types: list[str]
    is_active: bool
    failure_count: int
    last_delivery_at: datetime | None
    created_at: datetime


class WebhookIssued(WebhookRead):
    signing_secret: str


class PricingSimulationRequest(BaseModel):
    principal: Decimal = Field(gt=0)
    rate_percent: Decimal = Field(ge=0)
    term_months: int = Field(ge=1, le=120)
    processing_fee: Decimal = Field(default=Decimal("0"), ge=0)
    interest_method: str = "micro_loan"


class CompanyAssistantRequest(BaseModel):
    question: str = Field(min_length=3, max_length=500)

File: products/LoanHub/apps/frontend/api/companyOperatingSystem.ts
import { api } from "@/lib/api";

export type CompanyCapability = {
  number: number;
  key: string;
  name: string;
  implementation: string;
  status: string;
};

export type CompanyCommandDashboard = {
  generated_at: string;
  currency: string;
  portfolio: {
    loans_total: number;
    active_loans: number;
    overdue_loans: number;
    portfolio_balance: number;
    principal_originated: number;
    contractual_margin: number;
    overdue_scheduled_amount: number;
    expected_collections_30_days: number;
    par_1: number;
    par_7: number;
    par_30: number;
    par_60: number;
    par_90: number;
    branch_exposure: Record<string, number>;
  };
  treasury: {
    recorded_money_in: number;
    recorded_money_out: number;
    net_recorded_liquidity: number;
    entry_count: number;
  };
  collections: {
    open_cases: number;
    case_count: number;
    buckets: Record<string, number>;
    bucket_exposure: Record<string, number>;
    promise_to_pay: number;
    legal_handover: number;
    recovered_amount: number;
  };
  governance: {
    open_compliance_cases: number;
    high_risk_compliance_cases: number;
    open_reconciliation_exceptions: number;
    active_approval_workflows: number;
    unresolved_system_errors: number;
  };
  operations: {
    open_records: number;
    overdue_actions: number;
    by_module: Record<string, number>;
  };
  cash_flow: { successful_payment_volume: number; successful_payment_count: number };
  profitability: { contractual_margin: number; recoveries: number; note: string };
  liquidity_forecast: {
    recorded_net_liquidity: number;
    expected_collections_30_days: number;
    illustrative_30_day_position: number;
    note: string;
  };
  warnings: Array<{ level: string; title: string; detail: string }>;
};

export type OperatingRecord = {
  id: string;
  company_id: string;
  branch_id: string | null;
  module: string;
  record_type: string;
  reference: string;
  title: string;
  description: string | null;
  status: string;
  priority: string;
  borrower_id: string | null;
  loan_id: string | null;
  assigned_user_id: string | null;
  counterparty_name: string | null;
  amount: number | null;
  currency: string;
  due_at: string | null;
  data: Record<string, unknown>;
  tags: string[];
  is_archived: boolean;
  created_at: string;
  updated_at: string;
};

export type APIKeyRecord = {
  id: string;
  name: string;
  key_prefix: string;
  scopes: string[];
  allowed_ips: string[];
  expires_at: string | null;
  last_used_at: string | null;
  revoked_at: string | null;
  created_at: string;
};

export type WebhookRecord = {
  id: string;
  name: string;
  endpoint_url: string;
  secret_prefix: string;
  event_types: string[];
  is_active: boolean;
  failure_count: number;
  last_delivery_at: string | null;
  created_at: string;
};

export async function getCompanyCapabilities(): Promise<CompanyCapability[]> {
  return (await api.get<CompanyCapability[]>("/company-operating-system/capabilities")).data;
}

export async function getCompanyCommandDashboard(): Promise<CompanyCommandDashboard> {
  return (await api.get<CompanyCommandDashboard>("/company-operating-system/dashboard")).data;
}

export async function listOperatingRecords(module?: string): Promise<OperatingRecord[]> {
  return (await api.get<OperatingRecord[]>("/company-operating-system/records", { params: module ? { module } : undefined })).data;
}

export async function createOperatingRecord(payload: {
  module: string;
  record_type: string;
  title: string;
  description?: string;
  status?: string;
  priority?: string;
  counterparty_name?: string;
  amount?: number;
  due_at?: string;
  tags?: string[];
  data?: Record<string, unknown>;
}): Promise<OperatingRecord> {
  return (await api.post<OperatingRecord>("/company-operating-system/records", payload)).data;
}

export async function updateOperatingRecord(id: string, payload: Partial<OperatingRecord>): Promise<OperatingRecord> {
  return (await api.patch<OperatingRecord>(`/company-operating-system/records/${id}`, payload)).data;
}

export async function getCollectionsStrategy(): Promise<Record<string, unknown>> {
  return (await api.get<Record<string, unknown>>("/company-operating-system/collections/strategy")).data;
}

export async function getReconciliationSummary(): Promise<Record<string, unknown>> {
  return (await api.get<Record<string, unknown>>("/company-operating-system/reconciliation/summary")).data;
}

export async function simulateCompanyPricing(payload: {
  principal: number;
  rate_percent: number;
  term_months: number;
  processing_fee: number;
  interest_method: string;
}): Promise<Record<string, unknown>> {
  return (await api.post<Record<string, unknown>>("/company-operating-system/pricing/simulate", payload)).data;
}

export async function listCompanyApiKeys(): Promise<APIKeyRecord[]> {
  return (await api.get<APIKeyRecord[]>("/company-operating-system/api-keys")).data;
}

export async function createCompanyApiKey(payload: { name: string; scopes: string[] }): Promise<APIKeyRecord & { api_key: string }> {
  return (await api.post<APIKeyRecord & { api_key: string }>("/company-operating-system/api-keys", payload)).data;
}

export async function revokeCompanyApiKey(id: string): Promise<APIKeyRecord> {
  return (await api.post<APIKeyRecord>(`/company-operating-system/api-keys/${id}/revoke`)).data;
}

export async function listCompanyWebhooks(): Promise<WebhookRecord[]> {
  return (await api.get<WebhookRecord[]>("/company-operating-system/webhooks")).data;
}

export async function createCompanyWebhook(payload: { name: string; endpoint_url: string; event_types: string[] }): Promise<WebhookRecord & { signing_secret: string }> {
  return (await api.post<WebhookRecord & { signing_secret: string }>("/company-operating-system/webhooks", payload)).data;
}

export async function generateCompanyBoardPack(): Promise<Record<string, unknown>> {
  return (await api.post<Record<string, unknown>>("/company-operating-system/board-packs")).data;
}

export async function askCompanyDataAssistant(question: string): Promise<{
  answer: string;
  evidence: Record<string, unknown>;
  generated_at: string;
  mode: string;
  notice: string;
}> {
  return (await api.post("/company-operating-system/assistant", { question })).data;
}
